src/cargo/ops/cargo_compile/unit_generator.rs

```diff
@@ -521,16 +506,19 @@ Rustdoc did not scrape the following examples because they require dev-dependenc
                 let mut filters = String::new();
                 let mut miss_count = 0;
 
-                let mut append = |t: &FilterRule, s| {
-                    if let FilterRule::All = *t {
-                        miss_count += 1;
-                        filters.push_str(s);
-                    }
-                };
-
                 if all_targets {
                     filters.push_str(" `all-targets`");
                 } else {
+                    if *lib == LibRule::True {
+                        miss_count += 1;
+                        filters.push_str(" `lib`,");
+                    }
+                    let mut append = |t: &FilterRule, s| {
+                        if let FilterRule::All = *t {
+                            miss_count += 1;
+                            filters.push_str(s);
+                        }
+                    };
                     append(bins, " `bins`,");
                     append(tests, " `tests`,");
                     append(examples, " `examples`,");
```

Comment:
seems the earliest ancestor of the following out-of-diff warn is this: https://github.com/rust-lang/cargo/pull/9549/files#diff-98da1b66b532e50e9bca971b453ee7de96b17436de1d7ef3824c9888d55be9bbR1189-R1193 (around Jun 10, 2021)

```rust
                return shell.warn(format!(
                    "target {}{} specified, but no targets matched; this is a no-op",
                    if miss_count > 1 { "filters" } else { "filter" },
                    filters,
                ));
```